File: src/lib/deviner.ts
/**
 * « CHAP.CI ÉCRIT L'ANNONCE » — côté téléphone.
 *
 * Nouveauté n° 1 du 03/09/2026. Le vendeur prend la photo ; le titre, la
 * catégorie, la sous-catégorie, l'état et les caractéristiques se remplissent.
 * Ce fichier prépare ce qui part au serveur (`/annonce/deviner`) :
 *
 *  · la photo, RÉDUITE à 768 px : le moteur n'a pas besoin de plus pour
 *    reconnaître un téléphone, et chaque pixel envoyé coûte des jetons et du
 *    forfait — la photo pleine (1280 px) reste pour l'annonce ;
 *  · le catalogue des catégories, tel que le site le connaît. C'est le client
 *    qui le tient (`data/categories.ts`), pas le serveur : une seconde copie
 *    en PHP divergerait à la première catégorie ajoutée.
 *
 * Les sous-catégories n'ont pas d'identifiant à part : leur nom EST leur
 * identifiant dans les annonces (`listing.subcategory`). Le catalogue les
 * envoie donc telles quelles, et le serveur vérifie que la réponse du moteur
 * en fait partie.
 */
import { categories } from '../data/categories'
import { phpDeviner, phpDevinerDisponible, type CatalogueDeviner, type Devine } from './php'

export type { Devine }

export const devinerDisponible = phpDevinerDisponible

/** Le catalogue que le moteur lira — et contre lequel le serveur vérifiera sa réponse. */
export function catalogueDeviner(): CatalogueDeviner[] {
  return categories.map((c) => ({
    id: c.id,
    label: c.name,
    sous: c.subcategories.map((s) => ({ id: s, label: s })),
  }))
}

/** La photo réduite pour le moteur : 768 px de côté au plus, JPEG. */
export function reduirePourVision(dataUri: string, max = 768): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image()
    img.onload = () => {
      const k = Math.min(1, max / Math.max(img.naturalWidth, img.naturalHeight))
      const l = Math.max(1, Math.round(img.naturalWidth * k))
      const h = Math.max(1, Math.round(img.naturalHeight * k))
      const canvas = document.createElement('canvas')
      canvas.width = l
      canvas.height = h
      const ctx = canvas.getContext('2d')
      if (!ctx) { reject(new Error('canvas indisponible')); return }
      ctx.drawImage(img, 0, 0, l, h)
      try { resolve(canvas.toDataURL('image/jpeg', 0.8)) }
      catch (e) { reject(e instanceof Error ? e : new Error(String(e))) }
    }
    img.onerror = () => reject(new Error('photo illisible'))
    img.src = dataUri
  })
}

/** Demande au moteur de rédiger l'annonce à partir d'une photo (déjà en data-URI). */
export async function deviner(photo: string): Promise<Devine> {
  const petite = await reduirePourVision(photo)
  return phpDeviner(petite, catalogueDeviner())
}

File: flutter_app/lib/api/deviner.dart
// =============================================================================
//  « CHAP.CI ÉCRIT L'ANNONCE » — côté téléphone.
//
//  Nouveauté n° 1 du 03/09/2026, portée du site (`src/lib/deviner.ts`). Le
//  vendeur prend la photo ; le titre, la catégorie, la sous-catégorie, l'état
//  et les caractéristiques se remplissent. Ce fichier prépare ce qui part au
//  serveur (`POST /annonce/deviner`) :
//
//   · la photo, RÉDUITE à 768 px et recomprimée en JPEG : le moteur n'a pas
//     besoin de plus pour reconnaître un téléphone, et chaque pixel envoyé
//     coûte des jetons et du forfait — la photo pleine (1 600 px) reste pour
//     l'annonce. La réduction se fait dans un isolat (`compute`) : décoder et
//     réencoder une photo prend une demi-seconde, et l'écran ne doit pas
//     geler pendant ce temps ;
//   · le catalogue des catégories, tel que l'application le connaît. C'est
//     le client qui le tient (`data/categories.dart` + `registre.dart`), pas
//     le serveur : une seconde copie en PHP divergerait à la première
//     catégorie ajoutée.
//
//  Les sous-catégories n'ont pas d'identifiant à part : leur nom EST leur
//  identifiant dans les annonces (`listing.subcategory`). Le catalogue les
//  envoie donc telles quelles, et le serveur vérifie que la réponse du moteur
//  en fait partie — s'il invente, la case reste vide.
//
//  ⚠️ LA CLÉ DU MOTEUR N'EST PAS DANS L'APPLICATION. Elle est dans
//  `api/config.php`, sur le serveur, chez le Patron. L'application demande
//  d'abord si la fonction est allumée (`GET /annonce/deviner`) ; si non, rien
//  ne change à l'écran — le formulaire reste celui d'hier.
// =============================================================================
import 'dart:convert';
import 'dart:typed_data';

import 'package:flutter/foundation.dart' show compute;
import 'package:image/image.dart' as img;

import '../data/categories.dart';
import '../data/formulaires/registre.dart';
import 'api_client.dart';

/// Ce que le moteur a lu sur la photo. Tout peut être vide : on ne remplit
/// que ce qu'il a su dire, et jamais par-dessus ce que la personne a tapé.
class Devine {
  final String titre;
  final String description;
  final String categoryId;
  final String subcategory;

  /// 'neuf', 'occasion' ou '' (indécis).
  final String etat;

  /// Clé de champ (en minuscules) → valeur : marque, modèle, couleur…
  final Map<String, String> caracteristiques;

  /// 0 à 100. Sous 30, on n'applique rien : mieux vaut un formulaire vide
  /// qu'un formulaire faux.
  final int confiance;

  const Devine({
    required this.titre,
    required this.description,
    required this.categoryId,
    required this.subcategory,
    required this.etat,
    required this.caracteristiques,
    required this.confiance,
  });

  factory Devine.fromJson(Map<String, dynamic> j) {
    final car = <String, String>{};
    if (j['caracteristiques'] is Map) {
      (j['caracteristiques'] as Map).forEach((k, v) {
        final cle = k.toString().trim().toLowerCase();
        final val = (v ?? '').toString().trim();
        if (cle.isNotEmpty && val.isNotEmpty) car[cle] = val;
      });
    }
    return Devine(
      titre: (j['titre'] ?? '').toString().trim(),
      description: (j['description'] ?? '').toString().trim(),
      categoryId: (j['categoryId'] ?? '').toString().trim(),
      subcategory: (j['subcategory'] ?? '').toString().trim(),
      etat: (j['etat'] ?? '').toString().trim(),
      caracteristiques: car,
      confiance: (j['confiance'] is num) ? (j['confiance'] as num).toInt() : 0,
    );
  }
}

/// La fonction est-elle allumée sur le serveur (clé en place) ? `false` au
/// moindre doute : un réseau coupé ne doit pas déclencher d'attente.
Future<bool> devinerDisponible() async {
  try {
    final d = await ApiClient.instance.get('/annonce/deviner');
    return d is Map && d['disponible'] == true;
  } catch (_) {
    return false;
  }
}

/// Le catalogue que le moteur lira — et contre lequel le serveur vérifiera sa
/// réponse. Même forme que sur le site : `{id, label, sous: [{id, label}]}`.
List<Map<String, dynamic>> catalogueDeviner() => [
      for (final c in categories)
        {
          'id': c.id,
          'label': c.nom,
          'sous': [
            for (final s in sousDe(c.id)) {'id': s, 'label': s},
          ],
        },
    ];

/// Côté le plus long de la photo envoyée au moteur.
const int _coteMax = 768;

/// Réduit et recomprime — dans un isolat, voir l'en-tête. `null` si la photo
/// n'est pas décodable (format exotique) : on n'envoie alors rien.
Future<Uint8List?> reduirePourVision(Uint8List octets) =>
    compute(_reduire, octets);

Uint8List? _reduire(Uint8List octets) {
  try {
    var image = img.decodeImage(octets);
    if (image == null) return null;
    // Les photos de téléphone portent leur rotation dans l'EXIF : sans ce
    // redressement, le moteur lirait un frigo couché.
    image = img.bakeOrientation(image);
    final plusGrand = image.width > image.height ? image.width : image.height;
    if (plusGrand > _coteMax) {
      final k = _coteMax / plusGrand;
      image = img.copyResize(
        image,
        width: (image.width * k).round().clamp(1, _coteMax),
        height: (image.height * k).round().clamp(1, _coteMax),
        interpolation: img.Interpolation.average,
      );
    }
    return Uint8List.fromList(img.encodeJpg(image, quality: 80));
  } catch (_) {
    return null;
  }
}

/// Demande au moteur de rédiger l'annonce à partir d'une photo. Lève une
/// [ApiException] si le serveur refuse (quota du jour, photo refusée, moteur
/// éteint) — l'écran décide alors de se taire.
Future<Devine?> deviner(Uint8List photo) async {
  final petite = await reduirePourVision(photo);
  if (petite == null) return null;
  final d = await ApiClient.instance.post(
    '/annonce/deviner',
    {
      'image': 'data:image/jpeg;base64,${base64Encode(petite)}',
      'catalogue': catalogueDeviner(),
    },
    // Le serveur attend le moteur jusqu'à 60 s ; trois à dix en pratique.
    timeout: const Duration(seconds: 70),
  );
  if (d is Map<String, dynamic>) return Devine.fromJson(d);
  return null;
}
